(#Closes 468) Added Directive node separated from comments

LonelyCat124 opened the pull request:
Initial implementation of this feature.

Ready for a look from any of @arporter @sergisiso @hiker

LonelyCat124 commented:
Pushed some changes now to fix some docs issues and coverage updates.

arporter reviewed (changes requested):
Thanks Aidan. Everything looks sensible but I do have a couple questions about inheritance and valid locations for directives.
In fact, now that I write that, please could you add a new section to the documentation - probably just before https://fparser.readthedocs.io/en/latest/fparser2.html#preprocessing-directives.
I'm wondering whether it would be wise for us to make this behaviour configurable somehow, just in case interpreting things as directives causes problems for downstream applications. I guess we could do this in the same way as we do for the support for non-standard extensions.

LonelyCat124 commented on src/fparser/two/Fortran2003.py:
```diff
@@ -116,9 +116,86 @@
+            obj = object.__new__(cls)
+            obj.init(string)
+            return obj
+        elif isinstance(string, FortranReaderBase):
```
Tidied this up for the directive.

LonelyCat124 commented on src/fparser/two/tests/test_comments_and_directives.py:
```diff
@@ -409,3 +409,47 @@ def test_action_stmts():
+        if comment.items[0] != "":
+            assert comment.items[0] == "! A comment!"
+            comments = comments + 1
+    assert comments == 1
```
No - for some reason fparser makes a bunch of empty comments in various locations in this tree (this behaviour happened before my changes), so I need to count how many non-empty comments there are.

Do directives exist in fixed-format? I've never really written fixed-format source - i guess I need to check what `c    $` does? But i'm also not sure what its meant to do.

sergisiso commented:
@LonelyCat124 I know we talked about that but I forgot. What is the benefit of differentiating in fparser between Comments and Directives? Are you planning to parse the body of the directive in the future?

LonelyCat124 commented:
> @LonelyCat124 I know we talked about that but I forgot. What is the benefit of differentiating in fparser between Comments and Directives? Are you planning to parse the body of the directive in the future?

@sergisiso yes - its to separate directives from comments to make handling them in PSyclone better in the future (which I plan to continue down that avenue now). It also cleans up even the current implementation I think.

LonelyCat124 commented:
@arporter I think this is cleaned up and ready for another look now.

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -141,16 +142,20 @@ def __new__(cls, string: str | FortranReaderBase, parent_cls=None):
         if isinstance(string, readfortran.Comment):
-            # Directives must start with a $
-            if not string.comment[1:].lstrip().startswith("$"):
+            # Directives must start with a $ or be !dir$ or cdir$
```
Actually, I've changed my mind. Although the variations supported by the OMP spec are quite funky, they can only get interpreted as comments if we are handling fixed format. Therefore, I *think* if we have a comment that begins e.g. `*$omp` then it can only be an OMP directive. Therefore, I think all you need to do is extend the list of things you check for here. It might be nicer to have this list as a class or module property that is documented (e.g. saying which ones are omp, which ones are flang etc.). If you could then use Sphinx to import this documentation into the user guide then so much the better.

LonelyCat124 replied:
Added to the Directive class.

arporter reviewed (changes requested):
Thanks Aidan. Getting there now. I think we need a class member to hold all the various sentinels that we support. We also need a new section in the documentation (I suggested a location last time). Ideally we could generate some of that from the list of sentinels in the code using e.g. Sphinx `.. autoclass ....`

LonelyCat124 commented:
@arporter Should be ready for another look now I think.

arporter commented on doc/source/fparser2.rst:
```diff
@@ -495,6 +495,16 @@ file was found but would fail if the include file was not found::
+Compiler/OpenMP Directive support
+---------------------------------
+Most Fortran compilers support directives to enable compiler-specific
+functionality. Fparser will convert these into ``Directive`` nodes where
```
I suggest we probably want this to say "optionally" and specify how the user turns this on/off. I'll start a discussion on the PR to get people's views. (At bare minimum I think it already requires the user to have requested that comments be retained by the reader?)

arporter reviewed (changes requested):
Thanks Aidan, I think it's looking pretty good now. Unfortunately, I've realised we haven't tackled the slightly tricky problem of how much control we want to expose for this new feature.

My instinct is to say that it should be possible to turn this on/off in case it causes problems in a downstream application. What do you, @sergisiso and @hiker think?

hiker commented:
> Thanks Aidan, I think it's looking pretty good now. Unfortunately, I've realised we haven't tackled the slightly tricky problem of how much control we want to expose for this new feature.
> 
> My instinct is to say that it should be possible to turn this on/off in case it causes problems in a downstream application. What do you, @sergisiso and @hiker think?

I can't really judge the impact (e.g. I have a few tools in fparser, see example, could they be affected? I don't actually think so, so I am fine :) ). But I would stick with: fparser is a Fortran parser. And the directives are not part of the Fortran standard (i.e. they are comments, and should be treated as such ... i.e. being ignored) - by default. But then have options to treat them differently.

hiker commented on doc/source/fparser2.rst:
```diff
@@ -495,6 +495,16 @@ file was found but would fail if the include file was not found::
+
+The supported directives are those recognized by flang, ifx, ifort (``!dir$``),
+and gcc (``!gcc$``), as well as OpenMP directives (such as ``!$omp``
+or alternatives).
```
As mentioned before, an option imho. I believe I read a comment somewhere about comments: imho preserving comments should be different from directives, i.e. a user might want directives, but not comments(?)

Re OMP: that's actually (imho) a different thing: the omp conditional code converts to standard Fortran, I am not sure how this should get combined with other directives in general. As far as I can see, directives usually take the place where a statement could be. Not so for omp conditionals:
```
    x = x+                      &
!$    omp_get_thread_num() + &
    5
```

LonelyCat124 replied:
I added an option for this now.

LonelyCat124 commented on src/fparser/two/Fortran2003.py:
```diff
@@ -116,9 +116,104 @@
+    ]
+
+    @show_result
+    def __new__(cls, string: str | FortranReaderBase, parent_cls=None):
```
Yeah, fixed.

LonelyCat124 commented on src/fparser/two/Fortran2003.py:
```diff
@@ -116,9 +116,104 @@
+            if not (
+                any(
+                    [
+                        string.comment.lower().startswith(prefix)
```
Done

LonelyCat124 commented on src/fparser/two/Fortran2003.py:
```diff
@@ -194,20 +289,23 @@ def restore_reader(self, reader):
 
     """
-    obj = Comment(reader)
+    obj = Directive(reader)
```
Yeah - I added it now.

arporter commented on src/fparser/common/readfortran.py:
```diff
@@ -554,22 +554,30 @@ class FortranReaderBase:
+        mode: bool,
+        ignore_comments: bool,
+        include_omp_conditional_lines: bool = False,
+        process_directives: bool = False,
```
This new option needs associated tests please.

LonelyCat124 replied:
Added - notably it was only on an abstract class so I'm not sure how it was affecting actual code.

arporter reviewed (changes requested):
Thanks Aidan. I'm pretty happy now. Just a little bit of tidying remaining and I've realised we're missing unit tests for the new option to the Reader.

I've created a PSyclone PR (https://github.com/stfc/PSyclone/pull/3121) that adopts this branch of fparser and have set all the integration tests running, just to be on the safe side.

LonelyCat124 commented on src/fparser/common/readfortran.py:
```diff
@@ -554,22 +554,30 @@ class FortranReaderBase:
     :param mode: a FortranFormat object as returned by \
```
Done

LonelyCat124 commented on doc/source/fparser2.rst:
```diff
@@ -498,8 +498,11 @@ file was found but would fail if the include file was not found::
+``Directive`` nodes where possible. This option is ``process_directives``,
+and by default it is set to ``False``. For directives to be processed, the
+``ignore_comments`` must be also be ``False``, otherwise Fparser will ignore
+this option.
```
Added something small to show how its used, I'm not sure it needs a lot more as it would be covered by the general fparser usage sections I think. Happy to add more if you don't agree though.

LonelyCat124 commented:
@arporter Fixed it up now I think, I'd not understood how fparser2 actually works well enough to have it always working - I fixed it up now.

arporter commented:
Bad news; we appear to have broken Fab. The LFRic extraction test fails:
```
Building 'gungho_model_extract-fast-debug-mpif90-gfortran' took 0:00:27.212276
read_metric: recorded 526 metrics
matplotlib not installed, no metrics summary charts produced
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/apps/packages/python/3.13.7/lib/python3.13/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
                    ~~~~^^^^^^^^^^^^^^^
  File "/apps/packages/python/3.13.7/lib/python3.13/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/archive/psyclone-tests/action-runner-software/actions-runner/_work/PSyclone-mirror/PSyclone-mirror/.runner_venv/lib/python3.13/site-packages/fab/parse/fortran_common.py", line 116, in run
    if node_tree.content[0] is None:
       ~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```
Although the LFRic integration test itself passes. Probably we need @hiker's expertise here...

arporter commented on src/fparser/two/tests/test_comments_and_directives.py:
```diff
@@ -514,7 +517,9 @@ def test_all_directive_formats(directive, expected, free):
         source = source + "        end program foo"
 
-    reader = get_reader(source, isfree=free, ignore_comments=False)
+    reader = get_reader(
```
Please duplicate this bit and test that it is sufficient just to supply `process_directives=True` *without* setting `ignore_comments` (with an associated comment).

hiker commented:
> Bad news; we appear to have broken Fab. The LFRic extraction test fails:
> 
> ```
> Building 'gungho_model_extract-fast-debug-mpif90-gfortran' took 0:00:27.212276
> read_metric: recorded 526 metrics
> matplotlib not installed, no metrics summary charts produced
> multiprocessing.pool.RemoteTraceback: 
> """
> Traceback (most recent call last):
>   File "/apps/packages/python/3.13.7/lib/python3.13/multiprocessing/pool.py", line 125, in worker
>     result = (True, func(*args, **kwds))
>                     ~~~~^^^^^^^^^^^^^^^
>   File "/apps/packages/python/3.13.7/lib/python3.13/multiprocessing/pool.py", line 48, in mapstar
>     return list(map(*args))
>   File "/archive/psyclone-tests/action-runner-software/actions-runner/_work/PSyclone-mirror/PSyclone-mirror/.runner_venv/lib/python3.13/site-packages/fab/parse/fortran_common.py", line 116, in run
>     if node_tree.content[0] is None:
>        ~~~~~~~~~~~~~~~~~^^^
> IndexError: list index out of range
> ```
> 
> Although the LFRic integration test itself passes. Probably we need @hiker's expertise here... ...this might well not be the fault of this PR though - it could be due to some earlier change to fparser.

I think the PSyclone CI branch of Fab needs a patch that I had applied a while ago. Sorry, with UKMO moving slow, I have too many branches in progress. I will verify that.

hiker commented:
Confirmed, I needed to update the PSyclone CI branch of our lfric Fab build repo. That's done now, please try again.

arporter commented:
Thanks @hiker. I've just tried again and we now get a failure to find netcdf.mod:
```
Error compiling /archive/psyclone-tests/latest-run/extraction_test/lfric_apps/applications/gungho_model/fab_workspace/gungho_model_extract-fast-debug-mpif90-gfortran/build_output/psydata/extract_netcdf_base.f90:
Command failed with return code 1:
['/archive/psyclone-spack/psyclone-spack-Jun25/spack-repo/opt/spack/linux-rhel10-zen5/gcc-14.2.1/openmpi-5.0.6-7643sxs65dmw462noa4pwhem2tnpewyv/bin/mpif90', '-ffree-line-length-none', '-Wall', '-g', '-Werror=conversion', '-Werror=character-truncation', '-Werror=unused-value', '-Werror=tabs', '-std=f2008', '-fdefault-real-8', '-fdefault-double-8', '-fcheck=all', '-ffpe-trap=invalid,zero,overflow', '-Og', '-I', '/archive/psyclone-tests/latest-run/extraction_test/lfric_apps/applications/gungho_model/fab_workspace/gungho_model_extract-fast-debug-mpif90-gfortran/build_output', '-J', '/archive/psyclone-tests/latest-run/extraction_test/lfric_apps/applications/gungho_model/fab_workspace/gungho_model_extract-fast-debug-mpif90-gfortran/build_output', '-c', '-fopenmp', 'extract_netcdf_base.f90', '-o', '/archive/psyclone-tests/latest-run/extraction_test/lfric_apps/applications/gungho_model/fab_workspace/gungho_model_extract-fast-debug-mpif90-gfortran/build_output/_prebuild/extract_netcdf_base.2ca77f7a6.o']
extract_netcdf_base.f90:180:9:

  180 |     USE netcdf, ONLY: nf90_noerr, nf90_strerror
      |         1
Fatal Error: Cannot open module file ‘netcdf.mod’ for reading at (1): No such file or directory
```
Does that ring a bell for either yourself or @sergisiso ?

LonelyCat124 commented:
@arporter I think I've addressed the remaining comments now, and fixed up the confusing test to test what it actually wants to test. Once we have the issues with building the integration tests resolved this should be ok for another look.

sergisiso commented:
@hiker The include path to netcdf is in FFLAGS. Could it be that fab does not take the existing environment values, I noticed that in this line:
```
        ${LFRIC_CORE}/build.sh ./fab_gungho_model_extract.py   \
          --profile ${COMPILER_PROFILE} --nproc 4 -fc $LDMPI -ld $LDMPI  -cc $CC
```
you add them explictily but not FFLAGS.

hiker commented:
> @hiker The include path to netcdf is in FFLAGS. Could it be that fab does not take the existing environment values, I noticed that in this line:
> 
> ```
>         ${LFRIC_CORE}/build.sh ./fab_gungho_model_extract.py   \
>           --profile ${COMPILER_PROFILE} --nproc 4 -fc $LDMPI -ld $LDMPI  -cc $CC
> ```
> 
> you add them explictily but not FFLAGS.

Yes, indeed - after I updated the Fab version here, we now have an inconsistent set of lfric scripts and Fab in the PSyclone branch :(  In the new Fab version, support for environment variable has been removed, but it has been moved into the (new) FabBase class instead. BUT, the lfric script doesn't yet use the new FabBase class, it uses an older version (when it was still called BafBase), and the version used did not yet support FFLAGS.  Too many branches and repos (fab, lfric-baf, and then our LFRic CI build).

I've just tried to update the PSyclone CI branch for PSyclone's lfric builds again, it should now have the right version. But I don't have time to test, so feel free to ignore this (including failures in the extraction step for now), and I will verify and if necessary properly fix this on Monday.

Sorry for that mess, we are just in the middle of migrating scripts from Baf to Fab (and while this is technically only a name change, the code reviews do cause a few differences and conflicts), so this work is still sitting on a branch, not really ready to be merged in yet (i.e. has not been fully propagated through all three repos to check that all LFRic scripts still work). I shouldn't have updated fab itself (then again, not sure how I would have fixed the fparser change, which does require a small change in Fab ... maybe we should have created a custom Fab branch for you as well :) )

arporter commented on src/fparser/common/readfortran.py:
```diff
@@ -1776,6 +1783,10 @@ class FortranStringReader(FortranReaderBase):
+        specialised Directive nodes. Default is False (in which case
+        directives are left as comments). This option overrides the
+        ignore_comments input.
 
```
But the initial description of FortranReaderBase could still do with being updated.

LonelyCat124 replied:
Added the autoclass after the initial mention of `FortranReaderBase`.

arporter commented on src/fparser/common/tests/test_readfortran.py:
```diff
@@ -1730,3 +1730,22 @@ def test_conditional_include_omp_conditional_liness_free_format_multiple():
+
+
+def test_process_directives_option_read_fortran():
+    """Test handling of the process_directives option."""
```
Pls add a note to say that the tests for the actual functionality are in two/tests/test_comments_and_directives.py

LonelyCat124 replied:
Done.

arporter reviewed (changes requested):
Thanks Aidan (and Joerg). Very nearly there now - we'll ignore the problem with the LFRic extraction test as it's nothing to do with this PR.
There's just a tweak needed to the documentation and I've requested an extension to a docstring in the tests to clarify where the reader functionality is actually tested. (I think this again points to it being in the wrong place but that's not for this PR).

hiker commented:
FWIW, I've fixed the CI. Sorry, our fab and lfric branches are changing quit heavily atm, and we are struggling keeping everything after Fab *(lfric-Fab scripts, and our lfric CI) up-to-date.
